### PR TITLE
Adds StackTraceLogginBehavior to allow shouldLogStackTrace to log all…

### PR DESCRIPTION
**Pull request by Apjo**
… 4xx and 5xx errors if stackTraceLoggingBehavior FORCE_STACK_TRACE else fallback to default

### CODE REVIEW

File: backstopper-core/src/main/java/com/nike/backstopper/handler/ApiExceptionHandlerBase.java [hunk 1]
```diff
@@ -372,11 +376,20 @@ protected boolean shouldLogStackTrace(
         int statusCode, Collection<ApiError> filteredClientErrors, Throwable originalException,
         Throwable coreException, RequestInfoForLogging request
     ) {
-        // By default, 4xx should *not* log stack trace. Everything else should.
-        //noinspection RedundantIfStatement
-        if (statusCode >= 400 && statusCode < 500)
+        if(coreException instanceof ApiException) {
```
Comment (nicmunroe) [inline backstopper-core/src/main/java/com/nike/backstopper/handler/ApiExceptionHandlerBase.java]: I don't think this logic is correct. If `stackTraceLoggingBehavior` is null or `DEFER_TO_DEFAULT_BEHAVIOR` then we'd want the status code checks to be done. We only want to short circuit return true or false when `stackTraceLoggingBehavior` equals `FORCE_STACK_TRACE` or `FORCE_NO_STACK_TRACE`.

File: backstopper-core/src/test/java/com/nike/backstopper/handler/ApiExceptionHandlerBaseTest.java [hunk 2]
```diff
@@ -474,6 +475,109 @@ public void doHandleApiException_should_not_allow_error_uid_from_passed_in_extra
...
+
+        // when
+        boolean result = handler.shouldLogStackTrace(500, filteredClientErrors, originalException, apiException, reqMock);
+
+        // then
+        Assertions.assertThat(result).isEqualTo(true);
+    }
+
+    @Test
+    public void shouldLogStackTrace_should_not_log_all_errors_when_StackTraceLoggingBehavior_is_Default_for_an_ApiException() {
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/handler/ApiExceptionHandlerBaseTest.java]: This is not correct. If the StackTraceLoggingBehavior is default or null, then it should fallback to status code checks (i.e. no stack trace on 4xx, full stack trace for everything else).

File: backstopper-core/src/test/java/com/nike/backstopper/handler/ApiExceptionHandlerBaseTest.java [hunk 3]
```diff
@@ -474,6 +475,109 @@ public void doHandleApiException_should_not_allow_error_uid_from_passed_in_extra
...
+
+        // when
+        boolean result = handler.shouldLogStackTrace(400, filteredClientErrors, originalException, coreException, reqMock);
+
+        // then
+        Assertions.assertThat(result).isEqualTo(false);
+    }
+
+    @Test
+    public void shouldLogStackTrace_should_not_log_all_errors_when_StackTraceLoggingBehavior_is_FORCE_NO_STACK_TRACE() {
```
Comment (nicmunroe) [inline backstopper-core/src/test/java/com/nike/backstopper/handler/ApiExceptionHandlerBaseTest.java]: You're using a 400, so it wouldn't log the stack trace anyway. This should be using a 500 along with `FORCE_NO_STACK_TRACE` to prove that the StackTraceLoggingBehavior is doing something other than the default behavior.

File: backstopper-core/src/main/java/com/nike/backstopper/handler/ApiExceptionHandlerBase.java (hunk 1 above)
Comment (Apjo) [reply backstopper-core/src/main/java/com/nike/backstopper/handler/ApiExceptionHandlerBase.java]: i see, my bad, will fix this. Thanks!

**Comment by nicmunroe** [comment]
@Apjo The tests still need some cleanup, but I have some bikeshedding I want to do so I'll wrap it up in a separate PR. Thank you for the contribution!

**Pull request merged**